Add support for refreshToken() (#277)
knpuniversity/oauth2-client-bundle: README.md

The logged-in user will be an instance of `KnpU\OAuth2ClientBundle\Security\User\OAuthUser` and will
have the roles `ROLE_USER` and `ROLE_OAUTH_USER`.

## Refresh Tokens

If you store your access tokens for use at a later time, you should check that it's not expired before you use it again.
If it is expired, you can easily refresh it:

```php

> // Load the access token from wherever you have store it (eg. user session, database, etc)

Reviewer: ```suggestion
// Load the access token from persistence (eg. user session, database, etc)
```

Reviewer: Looking good!

The logged-in user will be an instance of `KnpU\OAuth2ClientBundle\Security\User\OAuthUser` and will
have the roles `ROLE_USER` and `ROLE_OAUTH_USER`.

## Storing and refreshing tokens

You have a couple of options to store access tokens for use at a later time:

1. Store the `AccessToken` object (eg. serializing into the session), this allows you to check expiry before refreshing
    ```php
    // Fetch and store the AccessToken
    $accessToken = $client->getAccessToken();
    $session->set('access_token', $accessToken);

    // Load the access token from the session, and refresh if required
    $accessToken = $session->get('access_token');

    if ($accessToken->hasExpired()) {

>         $accessToken = $client->refreshAccessToken($accessToken);

Reviewer 2: needs to be `$accessToken->getRefreshToken()`, right?

Author: Good catch!

        // Update the stored access token for next time
        $session->set('access_token', $accessToken);
    }
    ```

2. Store just the refresh token string (eg. in the dabatase `user.refresh_token`), this means you must always refresh
    ```php
    // Fetch the AccessToken and store the refresh token
    $accessToken = $client->getAccessToken();
    $user->setRefreshToken($accessToken->getRefreshToken());

>     $entityManager->flush();

Reviewer 2: You could ALSO store the `access_token` and expires in the database if you wanted to. Maybe we could make a note of that?

```php
// you can also store the access token and expiration and then avoid the refresh
// until the access token is actually expired
```

    // Get a new AccessToken from the refresh token, and store the new refresh token for next time
    $accessToken = $client->refreshAccessToken($user->getRefreshToken);
    $user->setRefreshToken($accessToken->getRefreshToken());
    $entityManager->flush();
```

Depending on your OAuth2 provider, you may need to pass some parameters when refreshing the token:

```php

> $accessToken = $client->refreshAccessToken($accessToken, ['scopes' => 'offline_access']);

Reviewer 2: I think they may also need to do this during the original redirect.
